lib/improver/spotdata/extract_data.py

```diff
@@ -0,0 +1,374 @@
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+"""Gridded data extraction for the Improver site specific process chain."""
+
+import numpy as np
+from iris.coords import AuxCoord, DimCoord
+from iris import Constraint
+from iris.cube import Cube
+from numpy.linalg import lstsq
+from improver.spotdata.common_functions import (nearest_n_neighbours,
+                                                node_edge_test,
+                                                isclose)
+from improver.spotdata.read_input import data_from_dictionary
+from improver.constants import (R_DRY_AIR,
+                                CP_DRY_AIR)
+
+
+class ExtractData(object):
+    """
+    A series of methods for extracting data from grid points to derive
+    diagnostic values at off grid positions.
+
+    """
+
+    def __init__(self, method='use_nearest'):
+        """
+        The class is called with the desired method to be used in extracting/
+        interpolating data to the site of interest from gridded data.
+
+        """
+
+        self.method = method
+
+    def process(self, cube, sites, neighbours, ancillary_data, additional_data,
+                no_neighbours=9):
+        """
+        Call the correct function to enact the method of data extraction
+        specified. This function also handles multiple timesteps, consolidating
+        the resulting data cubes into an Iris.CubeList.
+
+        Args:
+        -----
+        cube : iris.cube.Cube
+            Cube of diagnostic data from which to extract spotdata.
+
+        sites : dict
+            A dictionary containing the properties of spotdata sites.
+
+        neighbours : numpy.array
+            Array of neigbouring grid points that are associated with sites
+            in the SortedDictionary of sites.
+
+        additional_data : dict
+            A dictionary containing any supplmentary time varying diagnostics
+            that are needed for the selected extraction method.
+
+        ancillary_data : dict
+            A dictionary containing additional model data that is needed.
+            e.g. {'orography': <cube of orography>}
+
+        no_neighbours : int
+            Number of grid points about the site to consider when calculating a
+            local temperature lapse rate using orography variation.
+            e.g. consider a 5x5 grid of points -> no_neighbours = 25.
```

Comment:
Good point. What's the behaviour if you don't specify a value that is (odd number)^2 e.g. `no_neighbours = 7`?